core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -81,9 +82,9 @@ class TransactionStateManager(brokerId: Int,
   private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
 
   def enablePidExpiration() {
-    scheduler.startup()
-
-    // TODO: add transaction and pid expiration logic
+    if (!scheduler.isStarted)
+      scheduler.startup()
+    // TODO: add pid expiration logic
```

Comment:
We could just schedule the recurring expiration func here (see my previous comment). Note that since this scheduler only have one background thread hence all scheduled tasks are ordered by the calling order, it will make the follow-up locking easier as well.